### PR TITLE
Fix zeroFields setting when merging slices into maps

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
viper, by design, doesn't support [case sensitive keys](https://github.com/spf13/viper#does-viper-support-case-sensitive-keys) and that ends up lower casing all keys regardless of whether the target field is a map. Flyte config has been using a workaround where we set the config values (e.g. in yaml files) for maps as yaml arrays instead to preserve their case sensitivity. After [this change](https://github.com/flyteorg/flytestdlib/pull/81) that sets ZeroFields to true, the logic in mapstructure package breaks when doing the slice to map conversion. This PR adds an additional hook that does the conversion.

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/948

### CODE REVIEW

File: config/viper/viper.go [hunk 1]
```diff
@@ -165,6 +165,35 @@ func canGetElement(t reflect.Kind) bool {
...
+// reverting that process.
+func sliceToMapHook(f reflect.Kind, t reflect.Kind, data interface{}) (interface{}, error) {
+	// Only handle slice -> map conversion
+	if f == reflect.Slice && t == reflect.Map {
+		// this will be the target result
+		res := map[interface{}]interface{}{}
+		// It's safe to convert data into a slice since we did the type assertion above.
+		asSlice := data.([]interface{})
+		for _, item := range asSlice {
+			//
```
Comment (katrogan) [inline config/viper/viper.go:180]: nit: rm?

**Review by katrogan — DISMISSED** [review]

**Review by katrogan — APPROVED** [review]
